target/linux/ipq806x/patches-5.10/097-1-drivers-net-dsa-qca8k-add-initial-support-for-led-co.patch

```diff
@@ -0,0 +1,106 @@
+From c54093931c22a72cbe3bf66c4d5acbb3f8cef390 Mon Sep 17 00:00:00 2001
+From: Ansuel Smith <ansuelsmth@gmail.com>
+Date: Thu, 4 Mar 2021 10:03:32 +0100
+Subject: [PATCH 09/10] drivers: net: dsa: qca8k: add initial support for led
+ config
+
+Some device use a non standard configuration for the LED blink freq.
+Add support for this specific configuration. If needed other LED
+configuration can be added later.
+
+Signed-off-by: Ansuel Smith <ansuelsmth@gmail.com>
+---
+ drivers/net/dsa/qca8k.c | 55 +++++++++++++++++++++++++++++++++++++++++
+ drivers/net/dsa/qca8k.h | 10 ++++++++
+ 2 files changed, 65 insertions(+)
+
+--- a/drivers/net/dsa/qca8k.c
++++ b/drivers/net/dsa/qca8k.c
+@@ -721,6 +721,57 @@ qca8k_setup_mac_pwr_sel(struct qca8k_pri
+ }
+ 
+ static int
++qca8k_setup_led_ctrl(struct qca8k_priv *priv)
++{
++	struct device_node *node = priv->dev->of_node;
++	u32 mask, val, reg;
++	int ret, rule;
++	u8 led_hz;
++
++	ret = of_property_read_u8(node, "qca8k,led_blink_freq", &led_hz);
++	if (!ret) {
++		switch (led_hz) {
++		case 8: /* 8 Hz */
++			val = 0x0;
++			break;
++		case 4: /* 4 Hz */
++			val = 0x1;
++			break;
++		case 2: /* 2 Hz */
++			val = 0x2;
++			break;
++		case 0: /* auto (1000mbps 8Hz, 100mbps 4Hz, 10mbps 2Hz) */
++			val = 0x3;
++			break;
++		default:
++			dev_err(priv->dev, "Unsupported led blink freq");
++			return -EINVAL;
++		}
++
++		for (rule = 0; rule < QCA8K_LED_RULE_COUNT; rule++) {
++			/* RULE 1, 3, 5 requires an offset */
++			if (rule % 2 == 0) {
++				mask <<= 16;
++				val <<= 16;
```

Comment:
`val` should be reset to initial value before `if (rule % 2 == 0) {` in every loop.